Feature: add addressbook lookup command

boscovn opened the pull request:
Added address book lookup command option, if set either via flag or as an entry in the config file, this is how the execution will change:
- It will read the command's stdout and extract from there a map of names indexed by email address
- During mail parsing, when creating a new AddressData struct, it will check the address in said map
- If it finds a match, it will:
    - Set the AddressData.Name value
    - Skip not just adding the name to the list, but parsing it from the email alltogether
    - Delete the key from the addressbook match
- When returning an existing AddressData, instead of checking the addressbook, it checks the AddressData.Name variable.
- The AddressData.Name variable is once again checked before executing the function that checks for name in addresses to skip it if already set
- After outputting the classed data, it iterates over the remaining addressbook entries in the map (those that weren't deleted)
- It instantiates AddressData structs with its values and prints them according to the format

boscovn commented:
Messed up rebasing, fixing now

boscovn commented:
Done

ferdinandyb commented:
If I get this right, your goal is to be able to combine outputs from both m-r-a and some hand curated address book, right? Maybe you should take a look at this project: https://jasoncarloscox.com/creations/addr-book-combine/ as it does just that. I should probably mention that in the README somewhere.

boscovn commented:
That was what I intended to do. I wasn't aware that a similar solution already existed, will look deeper into it, thanks

boscovn commented:
I looked up addr-book-combine, the approach is somewhat different, it dynamically queries the addressbooks (or any other sources) instead of making a single command to get all results.
There is however some value in embedding a lookup command in this one, that is for choosing the curated name from the addressbook instead of determining it from the algorithm, note that the functionality of the overall program remains unmodified, the only thing that changes in the generated table is the names.
There is even a use case for embedding a lookup command in this one AND using addr-book-combine, when you want the generated table to have curated names but also want to combine it with other sources. Also, addr-book-combine could be the command embedded in this one if you want the addressbook that this program queries to be gathered from multiple sources.
I'm going to add an additional flag in my fork to determine if you want the addressbook to be used for querying names to append the unmatched contacts to the generated table or not, if you're interested in this addition I will reopen the PR when I implement it

ferdinandyb commented:
I've reopened the PR until we come to a conclusion! @jasonccox can you comment on the use case? Tbh, I haven't myself tried your `addr-book-combine` so I'm not quite sure of the possibilities. Also, previously I was musing about adding a config option to `m-r-a` where you could override the names figured out automatically, but I would understand the appeal of being able to use and existing and proper contact management software for that. 

On a more technical note: I won't have time for a couple days to look into anything, but before accepting a new PR I want to clear #4 (except for tagging a new release). I hope @jasonccox can answer soon, and I'll probably be able to get back to this around Thursday.

ferdinandyb commented:
@boscovn on a side note: I think the email you made the commit with is not the same as your github user's email, so you are not showing up as a contributor.

boscovn commented:
Thanks for the heads up, added my new mail to github.
Will make some changes to the way this is configured while you're afk

jasonccox commented:
Hey @boscovn, I believe addr-book-combine could mostly work for your use case and doesn't need to be as dynamic as my examples make it look. For example, you could do the following, assuming you use khard for your address book:

```
addr-book-combine \
  -c "khard email --remove-first-line --parsable" \
  -c "maildir-rank-addr --outputpath -" \
  >addr-book.tsv
```

Since `khard email` is invoked without any query, it would print the entire address book. As a result, you'd get all of the emails in your address book, sorted by their rank from maildir-rank-addr but with the name from the address book, as well as all of the other names and addresses found by maildir-rank-addr. (This assumes that `--outputpath -` will make maildir-rank-addr print its addresses to stdout, which may not be true (yet!).)

The one caveat is that, as you mentioned, this method will cause all of the addresses stored in khard to be output, including those that maildir-rank-addr doesn't find.

Your integrated solution sounds nice as well, and could potentially be more efficient since it only looks up addresses that maildir-rank-addr actually finds and skips parsing names that appear in the address book.

boscovn commented:
Hey @jasonccox , I've added a flag to skip adding unmatched addressbook results, which it does by default, so that caveat is now dealt with.
I've also added another check before appending a name the AddressData name list, which is to skip empty names.

boscovn commented:
@jasonccox Upon re-reading your message I realized I didn't fully address your comment, my modifications don't make the assumption that maildir-rank-addr outputs to stdout, it parses the stdout of khard email (or whatever lookup command you set it to, as long as the output is [address]\t[name]) and adds it to a map in memory, then as it goes through the emails what it actually skips is parsing the name whenever there's a match by address with this auxiliary map, and assigning it instead of appending it to a list of potential names. Once it's done it outputs according to the template as usual, and after that it (optionally) adds to the same file the names and addresses in your contacts that weren't matched to emails, again respecting the same go template though obviously the value of everything that isn't name or address will be empty

ferdinandyb commented:
@boscovn Sorry for taking a bit longer, things got a little hairy around here. I updated the readme and added the contributing guidelines. If the PR goes forward please update it accordingly, even github is pretty strict with commit title lengths it seems :)

![image](https://user-images.githubusercontent.com/6343487/215272061-efe6be54-021f-4f20-b8aa-439033403c6d.png)

ferdinandyb commented:
> @jasonccox Upon re-reading your message I realized I didn't fully address your comment, my modifications don't make the assumption that maildir-rank-addr outputs to stdout, it parses the stdout of khard email (or whatever lookup command you set it to, as long as the output is [address]\t[name]) and adds it to a map in memory, then as it goes through the emails what it actually skips is parsing the name whenever there's a match by address with this auxiliary map, and assigning it instead of appending it to a list of potential names. Once it's done it outputs according to the template as usual, and after that it (optionally) adds to the same file the names and addresses in your contacts that weren't matched to emails, again respecting the same go template though obviously the value of everything that isn't name or address will be empty

I'm also not quite sure what is going on :D But in general I think I like it that `mra` should be able to use say khard to overwrite names of any addresses from say khard. If I understand correctly what this PR would do is load up everything from say khard and replace the names found by mra with those from khard if they exist in khard. What happens with names that are in khard but not found by mra? I think those should be discarded since we don't really know how to rank them? Or just put them at the bottom?

ferdinandyb commented on README.md:
````diff
@@ -45,6 +46,7 @@ or systemd timer).
 Supported flags:
 
 ```
+      --addressbookLookupCommand string   optional command to query addresses from your addressbook
````
`--addr-book-cmd` would be better I think

ferdinandyb commented on maildir-rank-addr.go:
```diff
@@ -4,7 +4,8 @@ import ()
 func main() {
 	config := loadConfig()
-	data := walkMaildir(config.maildir, config.addresses, config.customFilters)
+    addressbook:= parseAddressbook(config.addressbookLookupCommand)
```
please also fix indentation!

boscovn commented:
> What happens with names that are in khard but not found by mra?

They are discarded by default or added at the bottom if stated. Will make those changes when I get back on my laptop

ferdinandyb commented:
> > What happens with names that are in khard but not found by mra?
> 
> They are discarded by default or added at the bottom if stated. Will make those changes when I get back on my laptop

sounds perfect, thanks!

ferdinandyb commented:
Thanks for the changes! Can you please rebase on master instead of merging the changes from there and squash as appropriate? I can't apply your patch as it is. Tbh I think the whole the PR can safely be one commit (see [commit section](https://github.com/ferdinandyb/maildir-rank-addr/blob/master/CONTRIBUTING.md) in contributing. And just in case run `gofumpt`. I should probably set up a CI ...

boscovn committed:
Ability to use external contact management software for assigning names
Currently names for addresses are assigned based on how common they are in the users email, but this might not be always preferable. Add the ability to override names calculated by maildir-rank-addr using an external address lookup command.

boscovn commented:
Done!

boscovn commented:
Typo in commit message

ferdinandyb commented:
Super @boscovn, thanks! If you're already in it, could you please add a two sentence explanation to the commit? i.e. something like this:

> Currently names for addresses are assigned based on how common they are in the users email, but this might not be always preferable. Add the ability to override names calculated by maildir-rank-addr using an external address lookup command.

(Sorry about the nitpicking, but it really does help keep things tidy and with figuring out issues later, and also helps people who just do a git pull in figuring out what happened in an update, and I don't know how I can edit the commit and still have github keep your contribution linked to you.)

See contrib:
> Use the body of the commit message to actually explain what your patch does and why it is useful. Even if your patch is a one line fix, the description is not limited in length and may span over multiple paragraphs. Use proper English syntax, grammar and punctuation.

ferdinandyb commented:
I'll also tag a new release once it's in for good measure.

ferdinandyb commented:
Cool, I just tested, now that I could `curl | git apply` the patch, works perfectly!

boscovn commented:
Done

Sent from Proton Mail mobile

-------- Original Message --------
On 30 Jan 2023, 22:20, Bence Ferdinandy wrote:

> @ferdinandyb approved this pull request.
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/ferdinandyb/maildir-rank-addr/pull/3#pullrequestreview-1276055452), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ACGESG2K7UGBQO2OFZXU4C3WVAWC5ANCNFSM6AAAAAAT2VSMIE).
> You are receiving this because you were mentioned.Message ID: ***@***.***>
